hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/contract/s3a/ITestS3AContractBulkDelete.java

```diff
@@ -0,0 +1,185 @@
+     */
+    private static final int DELETE_PAGE_SIZE = 20;
+
+    private final boolean enableMultiObjectDelete;
+
+    @Parameterized.Parameters(name = "enableMultiObjectDelete = {0}")
+    public static Iterable<Object[]> enableMultiObjectDelete() {
+        return Arrays.asList(new Object[][] {
+                {true},
+                {false}
+        });
+    }
+
+    public ITestS3AContractBulkDelete(boolean enableMultiObjectDelete) {
+        this.enableMultiObjectDelete = enableMultiObjectDelete;
+    }
+
+    @Override
+    protected Configuration createConfiguration() {
+        Configuration conf = super.createConfiguration();
+        S3ATestUtils.disableFilesystemCaching(conf);
+        S3ATestUtils.removeBaseAndBucketOverrides(conf,
+                Constants.BULK_DELETE_PAGE_SIZE);
+        conf.setInt(Constants.BULK_DELETE_PAGE_SIZE, DELETE_PAGE_SIZE);
+        conf.setBoolean(Constants.ENABLE_MULTI_DELETE, enableMultiObjectDelete);
+        return conf;
+    }
+
+    @Override
+    protected AbstractFSContract createContract(Configuration conf) {
+        return new S3AContract(createConfiguration());
+    }
+
+    @Override
+    public void validatePageSize() throws Exception {
+        int targetPageSize = DELETE_PAGE_SIZE;
+        if (!enableMultiObjectDelete) {
+            // if multi-object delete is disabled, page size should be 1.
+            targetPageSize = 1;
+        }
+        Assertions.assertThat(pageSize)
+                .describedAs("Page size should match the configured page size")
+                .isEqualTo(targetPageSize);
+    }
+
+    @Test
+    public void testBulkDeleteZeroPageSizePrecondition() throws Exception {
+        if(!enableMultiObjectDelete) {
+            // if multi-object delete is disabled, skip this test as
+            // page size is always 1.
+            skip("Multi-object delete is disabled");
+        }
+        Configuration conf = getContract().getConf();
+        conf.setInt(Constants.BULK_DELETE_PAGE_SIZE, 0);
+        Path testPath = path(getMethodName());
+        try (S3AFileSystem fs = S3ATestUtils.createTestFileSystem(conf)) {
+            intercept(IllegalArgumentException.class,
+                    () -> fs.createBulkDelete(testPath));
+        }
+    }
+
+    @Test
+    public void testPageSizeWhenMultiObjectsDisabled() throws Exception {
+        Configuration conf = getContract().getConf();
+        conf.setBoolean(Constants.ENABLE_MULTI_DELETE, false);
+        Path testPath = path(getMethodName());
+        try (S3AFileSystem fs = S3ATestUtils.createTestFileSystem(conf)) {
+            BulkDelete bulkDelete = fs.createBulkDelete(testPath);
+            Assertions.assertThat(bulkDelete.pageSize())
+                    .describedAs("Page size should be 1 when multi-object delete is disabled")
+                    .isEqualTo(1);
+        }
+    }
+
+    @Override
+    public void testDeletePathsDirectory() throws Exception {
```

Comment:
* skip adding this second path without multi-object delete, so we can verify L153 assertion always holds
* add a variant where he path deleted is a parent of the directory